virtual_ship/expedition/verify_schedule.py

```diff
@@ -0,0 +1,130 @@
+"""verify_schedule function and supporting classes."""
+
+from datetime import timedelta
+
+import pyproj
+from parcels import FieldSet
+
+from .input_data import InputData
+from .instrument_type import InstrumentType
+from .schedule import Schedule
+from .ship_config import ShipConfig
+from .waypoint import Waypoint
+
+
+def verify_schedule(
+    projection: pyproj.Geod,
+    ship_config: ShipConfig,
+    schedule: Schedule,
+    input_data: InputData,
+) -> None:
+    """
+    Verify waypoints are ordered by time, first waypoint has a start time, and that schedule is feasible in terms of time if no unexpected events happen.
+
+    :param projection: projection used to sail between waypoints.
+    :param ship_config: The cruise ship_configuration.
+    :param schedule: The schedule to verify.
+    :param input_data: Fieldsets that can be used to check for zero UV condition (is waypoint on land).
+    :raises PlanningError: If waypoints are not feasible or incorrect.
+    :raises ValueError: If there are no fieldsets in the ship_config, which are needed to verify all waypoints are on water.
+    """
+    if len(schedule.waypoints) == 0:
+        raise PlanningError("At least one waypoint must be provided.")
+
+    # check first waypoint has a time
+    if schedule.waypoints[0].time is None:
+        raise PlanningError("First waypoint must have a specified time.")
+
+    # check waypoint times are in ascending order
+    timed_waypoints = [wp for wp in schedule.waypoints if wp.time is not None]
+    if not all(
+        [
+            next.time >= cur.time
+            for cur, next in zip(timed_waypoints, timed_waypoints[1:])
+        ]
+    ):
+        raise PlanningError(
+            "Each waypoint should be timed after all previous waypoints"
+        )
+
+    # check if all waypoints are in water
+    # this is done by picking an arbitrary provided fieldset and checking if UV is not zero
+
+    # get all available fieldsets
+    available_fieldsets = [
+        fs
+        for fs in [
+            input_data.adcp_fieldset,
+            input_data.argo_float_fieldset,
+            input_data.ctd_fieldset,
+            input_data.drifter_fieldset,
+            input_data.ship_underwater_st_fieldset,
+        ]
+        if fs is not None
+    ]
+    # check if there are any fieldsets, else its an error
+    if len(available_fieldsets) == 0:
+        raise ValueError(
+            "No fieldsets provided to check if waypoints are on land. Assuming no provided fieldsets is an error."
```

Comment:
Vague error for users

Reply:
This will only happen if the user has zero instruments and no fieldsets are loaded